segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -385,58 +317,70 @@ private void logError(String segment, Throwable u) {
-     * If there is room for more data, we resume consuming from the socket.
-     */
-    private void pauseOrResumeReading() {
-        long bytesWaiting = getPendingBytes();
+    //endregion
 
-        if (bytesWaiting > HIGH_WATER_MARK) {
-            log.debug("Pausing writing from connection {}", connection);
-            connection.pauseReading();
+    //region WriterState
+
+    @ThreadSafe
+    private static class WriterState {
+        @Getter
+        private final Object ackLock = new Object();
+        @GuardedBy("this")
+        private long lastStoredEventNumber;
+        @GuardedBy("this")
+        private long lastAckedEventNumber;
+        @GuardedBy("this")
+        private long inFlightCount;
+
+        WriterState(long initialEventNumber) {
+            this.inFlightCount = 0;
+            this.lastStoredEventNumber = initialEventNumber;
+            this.lastAckedEventNumber = initialEventNumber;
         }
-        if (bytesWaiting < LOW_WATER_MARK) {
-            log.trace("Resuming writing from connection {}", connection);
-            connection.resumeReading();
+
+        /**
+         * Invoked when a new append is initiated.
+         *
+         * @param eventNumber The Append's Event Number.
+         * @return The previously attempted Event Number.
+         */
+        synchronized long beginAppend(long eventNumber) {
+            long previousEventNumber = this.lastStoredEventNumber;
+            Preconditions.checkState(eventNumber >= previousEventNumber, "Event was already appended.");
+            this.lastStoredEventNumber = eventNumber;
+            this.inFlightCount++;
+            return previousEventNumber;
         }
-    }
 
-    /**
-     * return the remaining bytes to be consumed.
-     */
-    protected long getPendingBytes() {
-        return pendingBytes.get();
-    }
+        /**
+         * Invoked when a conditional append has failed due to {@link BadOffsetException}. If no more appends are in the
+         * pipeline, then the Last Stored Event Number is reverted to the Last (Successfully) Acked Event Number.
+         */
+        synchronized void conditionalAppendFailed() {
+            this.inFlightCount--;
+            if (this.inFlightCount == 0) {
+                this.lastStoredEventNumber = this.lastAckedEventNumber;
+            }
+        }
 
-    /**
-     * set the remaining bytes to be consumed.
-     * Make sure that negative value is not set.
-     *
-     * @param val  New value to set.
-     */
-    protected void setPendingBytes(long val) {
-        pendingBytes.set(Math.max(val, 0));
-    }
+        /**
+         * Invoked when an append has been successfully stored.
+         *
+         * @param eventNumber The Append's Event Number.
+         * @return The last successful Append's Event Number (prior to this one).
+         */
+        synchronized long appendSuccessful(long eventNumber) {
```

Comment:
This is subject to some very complicated race conditions. It looks like the Max takes care of this in conjunction with how it is called. But this should be called out in the javadoc, and a justification for correctness can be provided.

Reply:
Ok